### DISCUSSION TITLE
Local in outer scope defined after an inner block containing same-name local

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5757] [Q&A]
```cs
if(condition)
{
  string s = "abc";
  // ...
  return;
}
string s = "def"; // this is error
```
But should it be? Why can't we make that legal? There is no confusion here - the "abc" `s` is done with and not accessible anymore. 
Currently, I end up having to do this:
```cs
string s;
if(condition)
{
  s = "abc";
  // ...
  return;
}
s = "def";
```
The problem here is it can create a hidden bug:
```cs
string s;
if(condition)
{
  s = "abc";
  // ... - but no return
}
s = "def"; // now we're re-assigning a variable which was already set - is it desirable?
```
Whereas if we made the change:
```cs
if(condition)
{
  string s = "abc";
  // ...
}
string s = "def"; // we know this is a different local and therefore there is no confusion
```

**Comment by CyrusNajmabadi** [comment]
> But should it be? Why can't we make that legal? 

THis continues to find common coding mistakes for me.  Where i actually intend there to be one variable, and i don't realize there are two in these scopes like this.

> s = "def"; // now we're re-assigning a variable which was already set - is it desirable?

If the variable is set, and the not read before being written again, then we report an analyzer message in roslyn.  If it is read, then it's entirely dependent on the domain if it's intended for the overwrite to be ok or not.
